onnxruntime/core/providers/coreml/builders/helper.cc

```diff
@@ -122,5 +127,41 @@ std::vector<std::vector<NodeIndex>> GetSupportedNodes(const GraphViewer& graph_v
   return supported_node_vecs;
 }
 
+bool HasNeuralEngine(const logging::Logger& logger) {
+  bool has_neural_engine = false;
+
+#ifdef __APPLE__
+  struct utsname system_info;
+  uname(&system_info);
+  LOGS(logger, VERBOSE) << "Current Apple hardware info: " << system_info.machine;
+
+#if TARGET_OS_IPHONE
+  // utsname.machine has device identifier. For example, identifier for iPhone Xs is "iPhone11,2".
+  // Since Neural Engine is only available for use on A12 and later, major device version in the
+  // identifier is checked for these models:
+  // A12: iPhone XS (11,2), iPad Mini - 5th Gen (11,1)
+  // A12X: iPad Pro - 3rd Gen (8,1)
+  // For more information, see https://www.theiphonewiki.com/wiki/Models
+  if (strncmp("iPad", system_info.machine, 4) == 0) {
```

Comment:
is system_info.machine guaranteed to be at least 4 characters long?

Reply:
It should be based on all the current  system_info.machine for iOS devices,
But I think it's good idea to check the length anyway